File: majority-element-2/majority-element-2.py
class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        m1, m2, m1Count, m2Count = 0, 0, 0, 0

        for num in nums:
            if m1 == num:
                m1Count += 1
            elif m2 == num:
                m2Count += 1
            elif m1Count == 0:
                m1 = num
                m1Count = 1
            elif m2Count == 0:
                m2 = num
                m2Count = 1
            else:
                m1Count -= 1
                m2Count -= 1

        m1Count = m2Count = 0
        for num in nums:
            if m1 == num:
                m1Count += 1
            elif m2 == num:
                m2Count += 1

        res = []
        if m1Count > (len(nums) / 3):
            res.append(m1)
        if m2Count > (len(nums) / 3):
            res.append(m2)
        return res
                
sln = Solution()
print sln.majorityElement([1, 2, 3, 4, 5, 5, 5, 5, 5, 5, 5])

File: io/github/zhengyhn/leetcode/majority_element_2/majority-element-2.c
#include <stdio.h>
#include <stdlib.h>

/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* majorityElement(int* nums, int numsSize, int* returnSize) {
  int m1, m2, m1_count, m2_count;

  m1 = m2 = m1_count = m2_count = 0;

  for (int i = 0; i < numsSize; ++i) {
    if (m1 == nums[i]) {
      ++m1_count;
    } else if (m2 == nums[i]) {
      ++m2_count;
    } else if (m1_count == 0) {
      m1 = nums[i];
      m1_count = 1;
    } else if (m2_count == 0) {
      m2 = nums[i];
      m2_count = 1;
    } else {
      --m1_count;
      --m2_count;
    }
  }

  m1_count = m2_count = 0;
  for (int i = 0; i < numsSize; ++i) {
    if (nums[i] == m1) {
      ++m1_count;
    } else if (nums[i] == m2) {
      ++m2_count;
    }
  }

  int *res2 = (int *)malloc(2 * sizeof(int));
  int *res1 = (int *)malloc(sizeof(int));

  *returnSize = 0;
  if (m1_count > (numsSize / 3)) {
    res1[*returnSize] = res2[*returnSize] = m1;
    ++(*returnSize);
  }
  if (m2_count > (numsSize / 3)) {
    res1[*returnSize] = res2[*returnSize] = m2;
    ++(*returnSize);
  }
  if (*returnSize == 1) {
    free(res2);
    return res1;
  } else {
    free(res1);
    return res2;
  }
}

int main(void)
{
  int arr[10] = {3, 2, 3, 5, 5, 5, 5, 3, 2, 2};
  int size;

  int *res = majorityElement(arr, 10, &size);
  for (int i = 0; i < size; ++i) {
    printf("%d\n", res[i]);
  }
  if (res != NULL) {
    free(res);
  }
  return 0;
}
